### PR TITLE
speedups using hashset lookup and caching edges()

### PR DESCRIPTION
**Pull request by wynged**
BACKGROUND:
- Roof analysis function was running a bit slow on a mesh which lots of triangles.

DESCRIPTION:
- Use hashsets for testing if a certain point, line, or triangle exists in a list.  By creating reasonable HashCodes that relate to the geometry of the objects we can avoid having to do the actual geometric calculations each time we want to check if somethign already exists.
- upgrade to elements 0.9.1
- Cache the `Edges()` of the mesh, so that we can avoid recomputing them if they have not changed.  I have a couple checks for cache invalidation, but it might not be comprehensive.

TESTING:
- I think working with the RoofAnalysis function using a local version of GeometryEx is probably the best test at the moment.
  
FUTURE WORK:
- Remove the Comparers.cs file which contains comparers that should make it into Elements by version 0.9.2
- There might be a code logic improvement that could be made, but for now data structure speedups have gotten us the speed we need.

COMMENTS:
- Any other notes.

---
This change is [Reviewable](https://reviewable.io/reviews/hypar-io/geometryex/9)

### CODE REVIEW

File: GeometryEx/Comparers.cs [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        /// </summary>
+        public Vector3Comparer(double precision = Vector3.EPSILON)
+        {
+            _precision = precision;
+        }
+
+        /// <summary>
+        /// Are the two Vector3 geometrically equal within the tolerance.
+        /// </summary>
+        public override bool Equals(Vector3 x, Vector3 y)
```
Comment (anthonyhauck) [inline GeometryEx/Comparers.cs:42]: My attention kept tripping over whole vectors being compared with variables named 'x' and 'y'. Maybe rename them to 'a' and 'b' to avoid the momentary mental hiccup at wondering why the vectors x and y components are being compared (not what's happening of course, but it's an avoidable mental stumble) ?

File: GeometryEx/Comparers.cs [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    public class TriangleComparer : IEqualityComparer<Triangle>
+    {
+        private bool _rotationIndependent = false;
+        private double _precision = Vector3.EPSILON;
+        public TriangleComparer(bool rotationIndependent = false, double precision = Vector3.EPSILON)
+        {
+            _rotationIndependent = rotationIndependent;
+            _precision = precision;
+        }
+        public bool Equals(Triangle x, Triangle y)
```
Comment (anthonyhauck) [inline GeometryEx/Comparers.cs:98]: As above, and for Lines.

File: GeometryEx/MeshEx.cs [hunk 3]
```diff
@@ -25,7 +26,10 @@ public static int AddTriangles(this Mesh mesh, List<Triangle> triangles)
                     continue;
                 }
                 mesh.AddTriangle(triangle);
-                triangle.Vertices.ToList().ForEach(v => mesh.AddVertex(v));
```
Comment (anthonyhauck) [inline GeometryEx/MeshEx.cs:28]: Why this change? Too confusing when compacted like the original?

**Review by anthonyhauck — CHANGES_REQUESTED** [review]
A couple of minor points for clarity, one question. The Hashset operations are cool, I was unfamiliar with these techniques.

**Review by wynged — COMMENTED** [review]
I think the "why this change" is about this line.  


This is a very minor speed improvement, because when you call ToList() you're technically making another list in memory, which we don't really need in this case.  I reverted it to be consistent with the other code as this is not a speed bottle neck.

> *[Reviewable](https://reviewable.io/reviews/hypar-io/geometryex/9#-:-MfnwXTq1Dn5pkO-FVCi:b7up46b)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @anthonyhauck)

___
*[GeometryEx/Comparers.cs, line 42 at r1](https://reviewable.io/reviews/hypar-io/geometryex/9#-MfnvgZggBAyVs0xRRsZ-r1-42:-Mfnvwmk-IwI98YoWSEe:b-1mkkpc) ([raw file](https://github.com/hypar-io/geometryex/blob/1f9a8b72768790180988deba158e2641b5c60929/GeometryEx/Comparers.cs#L42)):*
<details><summary><i>Previously, anthonyhauck (Anthony Hauck) wrote…</i></summary><blockquote>

My attention kept tripping over whole vectors being compared within variables named 'x' and 'y'. Maybe rename them to 'a' and 'b' to avoid the momentary mental hiccup at wondering why the vectors x and y components are being compared (not what's happening of course, but it's an avoidable mental stumble) ?
</blockquote></details>

done. That's a good call for sure.
___
*[GeometryEx/Comparers.cs, line 98 at r1](https://reviewable.io/reviews/hypar-io/geometryex/9#-MfnvgZggBAyVs0xRRsZ-r1-98:-Mfnw5iiBxSsQHLv8rFM:b-896fix) ([raw file](https://github.com/hypar-io/geometryex/blob/1f9a8b72768790180988deba158e2641b5c60929/GeometryEx/Comparers.cs#L98)):*
<details><summary><i>Previously, anthonyhauck (Anthony Hauck) wrote…</i></summary><blockquote>

As above, and for Lines.
</blockquote></details>

Done.
